File: C++/26-二叉搜索树与双向链表.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution
{
public:
    TreeNode *Convert(TreeNode *pRootOfTree)
    {
        if(pRootOfTree == nullptr)
            return nullptr;
        TreeNode* pre = nullptr;
        helper(pRootOfTree, pre);
        TreeNode* res = pRootOfTree;
        while(res->left)
            res = res->left;
        return res;
    }

    void helper(TreeNode *cur, TreeNode *&pre)
    {
        if(cur == nullptr)
            return;
        helper(cur->left, pre);
        cur->left = pre;
        if(pre)
            pre->right = cur;
        pre = cur;
        helper(cur->right, pre);
    }
};

File: Python/26-二叉搜索树与双向链表.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def Convert(self, pRootOfTree):
        # write code here
        if not pRootOfTree:
            return None
        pre = [None]
        self.helper(pRootOfTree, pre)
        res = pRootOfTree
        while res.left:
            res = res.left
        return res

    def helper(self, cur, pre):
        if not cur:
            return
        self.helper(cur.left, pre)
        cur.left = pre[0]
        if pre[0]:
            pre[0].right = cur
        pre[0] = cur
        self.helper(cur.right, pre)
